[ao][sparsity] Support for sparsifying data operations on raw torch tensors.

macandro96 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79535
* #79534
* #79254
* #79253
* #79427
* __->__ #79252
* #79251

The users can now pass in raw torch tensors and the base class handles all the parametrizations and masking

Example -
    >>> data_list = [('tensor_1', torch.randn(3,3)), ('tensor_2', torch.randn(4,4))]
    >>> defaults = {'sparsity_level': 0.7}
    >>> sparsifier = DerivedDataSparsifier(data_list = data_list, **defaults) # Some sparsifier that inherits BaseDataSparsifier
    >>> new_tensor_to_add = {'name': 'tensor_3', 'data': torch.randn(5,5), 'sparsity_level': 0.3}
    >>> sparsifier.add_data(**new_tensor_to_add)
    >>> # tensor_1 and tensor_2 will have sparsity_level of 0.7 but tensor_3 will have sparsity_level=0.3

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

Differential Revision: [D37164121](https://our.internmc.facebook.com/intern/diff/D37164121)

Commit:
[ao][sparsity] Support for torch tensors

The users can now pass in raw torch tensors and the base class handles all the parametrizations and masking

Example -
    >>> data_list = [('tensor_1', torch.randn(3,3)), ('tensor_2', torch.randn(4,4))]
    >>> defaults = {'sparsity_level': 0.7}
    >>> sparsifier = DerivedDataSparsifier(data_list = data_list, **defaults) # Some sparsifier that inherits BaseDataSparsifier
    >>> new_tensor_to_add = {'name': 'tensor_3', 'data': torch.randn(5,5), 'sparsity_level': 0.3}
    >>> sparsifier.add_data(**new_tensor_to_add)
    >>> # tensor_1 and tensor_2 will have sparsity_level of 0.7 but tensor_3 will have sparsity_level=0.3

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

[ghstack-poisoned]

HDCharles commented on torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:
```diff
@@ -38,39 +56,198 @@ def __init__(self, data_list: Optional[List[Tuple[str, Any]]] = None, **defaults
+            raise ValueError("data with specified name does not exist")
+        return self.state[name]['mask']
 
     def squash_mask(self, *args, **kwargs):
```
should probably leave the kwargs in the function definition rather than just using kwargs, 

I believe if you do something like def squash_mask(self, *args, leave_parametrize=True, **kwargs) it will work.

macandro96 replied:
got it

HDCharles reviewed (approved):
looks good aside from those comments, not going to block though

macandro96 commented on torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:
```diff
@@ -38,39 +56,198 @@ def __init__(self, data_list: Optional[List[Tuple[str, Any]]] = None, **defaults
+        assert type(data) in SUPPORTED_TYPES, \
+            f'data type unsupported for {name}'
+
+        if name in self.state:
```
Yep - can do. Will make changes.

z-a-f commented:
More descriptive name for the PR title is needed. `Support for toech tensors` doesn't tell a lot about where is this support added

z-a-f commented on torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:
```diff
@@ -38,39 +56,198 @@ def __init__(self, data_list: Optional[List[Tuple[str, Any]]] = None, **defaults
+        Args:
+        * state_dict - the dictionary that to which the current sparsifier needs to be restored to
+        * strict - If True - the sparsifier is reset and is restored exactly to the state in state_dict.
+            If False - the current sparsifier is not reset before loading the state_dict i.e. data added
+            before loading the state_dict is not erased.
+        """
+        states = copy.deepcopy(state_dict['state'])
+        data_groups = copy.deepcopy(state_dict['data_groups'])
+        container_state_dict = copy.deepcopy(state_dict['_container'])
+        if strict:
+            # if strict load -> then reset container
+            self._container = Container()
 
-    def load_state_dict(self, state_dict):
-        pass
+        self._load_container_from_state(states, data_groups, container_state_dict)
+
+        if not strict:
+            states.update(self.state)
+            data_groups.update(self.data_groups)
+
+        self.__setstate__({'state': states, 'data_groups': data_groups})
 
     def __setstate__(self, state):
-        pass
+        if '_container' in state:  # If container object is in state then load model
+            container_dict = state.pop('_container')
+            self._container = Container()
+            self._load_container_from_state(state['state'], state['data_groups'], container_dict)
+
+        self.__dict__.update(state)
 
     def __getstate__(self):
-        pass
+        return {
+            'defaults': self.defaults,
+            'state': self.state,
+            'data_groups': self.data_groups,
+            '_container': self._container.state_dict()
+        }
```
Let's split out the serialization routines into a separate diff -- there might be more focusd comments that way, and there might be some tests that we would apply

macandro96 replied:
Oh I thought I did. I seemed to have amended the commit by mistake. Let me create a separate diff.

macandro96 replied:
done

z-a-f commented on torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:
```diff
@@ -38,39 +56,198 @@ def __init__(self, data_list: Optional[List[Tuple[str, Any]]] = None, **defaults
         Note: The container model is private to the BaseDataSparsifier class
         """
-        pass
+        assert type(data) in SUPPORTED_TYPES, \
```
nit: `assert isinstance(data, SUPPORTED_TYPES)`

z-a-f commented on torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:
```diff
@@ -38,39 +56,198 @@ def __init__(self, data_list: Optional[List[Tuple[str, Any]]] = None, **defaults
-    def get_data(self, name: str):
-        pass
+        if not return_sparsified:
+            if not parametrize.is_parametrized(self._container, name):
+                raise ValueError("mask squashed - original mask value does not exist")
```
nit: If it's not parametrized, and the assumption is that the squash_mask was already called, shouldn't this just return the tensor from the container? If the intention is to return the parametrized vs. original, I think a better name for the argument shoud be something like

```
def get_data(self, name: str, return_original: bool = True):
```

macandro96 replied:
makes sense.

z-a-f commented on torch/ao/sparsity/experimental/data_sparsifier/base_data_sparsifier.py:
```diff
@@ -38,39 +56,198 @@ def __init__(self, data_list: Optional[List[Tuple[str, Any]]] = None, **defaults
-        pass
+        if '_container' in state:  # If container object is in state then load model
+            container_dict = state.pop('_container')
+            self._container = Container()
```
Is there a reason why a new instance of the container is created, instead of applying the state to the existing container?

macandro96 replied:
The reason is that to do a load_state_dict() on an nn module, it is essential that the model parameters already exist in the container. Also, while unpickling (or deserializing) I don't think the constructor is called - hence we will have to create a new container object.

Commit:
Update on "[ao][sparsity] Support for torch tensors"


The users can now pass in raw torch tensors and the base class handles all the parametrizations and masking

Example -
    >>> data_list = [('tensor_1', torch.randn(3,3)), ('tensor_2', torch.randn(4,4))]
    >>> defaults = {'sparsity_level': 0.7}
    >>> sparsifier = DerivedDataSparsifier(data_list = data_list, **defaults) # Some sparsifier that inherits BaseDataSparsifier
    >>> new_tensor_to_add = {'name': 'tensor_3', 'data': torch.randn(5,5), 'sparsity_level': 0.3}
    >>> sparsifier.add_data(**new_tensor_to_add)
    >>> # tensor_1 and tensor_2 will have sparsity_level of 0.7 but tensor_3 will have sparsity_level=0.3

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Support for torch tensors"


The users can now pass in raw torch tensors and the base class handles all the parametrizations and masking

Example -
    >>> data_list = [('tensor_1', torch.randn(3,3)), ('tensor_2', torch.randn(4,4))]
    >>> defaults = {'sparsity_level': 0.7}
    >>> sparsifier = DerivedDataSparsifier(data_list = data_list, **defaults) # Some sparsifier that inherits BaseDataSparsifier
    >>> new_tensor_to_add = {'name': 'tensor_3', 'data': torch.randn(5,5), 'sparsity_level': 0.3}
    >>> sparsifier.add_data(**new_tensor_to_add)
    >>> # tensor_1 and tensor_2 will have sparsity_level of 0.7 but tensor_3 will have sparsity_level=0.3

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Support for torch tensors"


The users can now pass in raw torch tensors and the base class handles all the parametrizations and masking

Example -
    >>> data_list = [('tensor_1', torch.randn(3,3)), ('tensor_2', torch.randn(4,4))]
    >>> defaults = {'sparsity_level': 0.7}
    >>> sparsifier = DerivedDataSparsifier(data_list = data_list, **defaults) # Some sparsifier that inherits BaseDataSparsifier
    >>> new_tensor_to_add = {'name': 'tensor_3', 'data': torch.randn(5,5), 'sparsity_level': 0.3}
    >>> sparsifier.add_data(**new_tensor_to_add)
    >>> # tensor_1 and tensor_2 will have sparsity_level of 0.7 but tensor_3 will have sparsity_level=0.3

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

[ghstack-poisoned]

HDCharles commented on test/ao/sparsity/test_data_sparsifier.py:
```diff
@@ -2,12 +2,152 @@
+        linear_state['step_count'] = linear_state.get('step_count', 0) + 1
+
+
+class TestBaseDataSparsiferType(TestCase):
```
I only just now understood whats going on with this test class, I hadn't realized before that this was a helper class for the actual test class TestBaseDataSparsifier. I'm not understanding the name here, feels like it should be something like TestBaseDataSparsifier(Generator/Helper/Runner) or something like that and it should have a comment similar to the comment for the base class describing this.

macandro96 replied:
The helper class could take in data of any time and then run tests on them, so I had suffixed with Type. Renamed it to TestBaseDataSparsifierRunner as it seems to be a better name. Added comments as well.

Commit:
Update on "[ao][sparsity] Support for sparsifying data operations on raw torch tensors."


The users can now pass in raw torch tensors and the base class handles all the parametrizations and masking

Example -
    >>> data_list = [('tensor_1', torch.randn(3,3)), ('tensor_2', torch.randn(4,4))]
    >>> defaults = {'sparsity_level': 0.7}
    >>> sparsifier = DerivedDataSparsifier(data_list = data_list, **defaults) # Some sparsifier that inherits BaseDataSparsifier
    >>> new_tensor_to_add = {'name': 'tensor_3', 'data': torch.randn(5,5), 'sparsity_level': 0.3}
    >>> sparsifier.add_data(**new_tensor_to_add)
    >>> # tensor_1 and tensor_2 will have sparsity_level of 0.7 but tensor_3 will have sparsity_level=0.3

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Support for sparsifying data operations on raw torch tensors."


The users can now pass in raw torch tensors and the base class handles all the parametrizations and masking

Example -
    >>> data_list = [('tensor_1', torch.randn(3,3)), ('tensor_2', torch.randn(4,4))]
    >>> defaults = {'sparsity_level': 0.7}
    >>> sparsifier = DerivedDataSparsifier(data_list = data_list, **defaults) # Some sparsifier that inherits BaseDataSparsifier
    >>> new_tensor_to_add = {'name': 'tensor_3', 'data': torch.randn(5,5), 'sparsity_level': 0.3}
    >>> sparsifier.add_data(**new_tensor_to_add)
    >>> # tensor_1 and tensor_2 will have sparsity_level of 0.7 but tensor_3 will have sparsity_level=0.3

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

[ghstack-poisoned]

z-a-f reviewed (approved):
lgtm

Commit:
Update on "[ao][sparsity] Support for sparsifying data operations on raw torch tensors."


The users can now pass in raw torch tensors and the base class handles all the parametrizations and masking

Example -
    >>> data_list = [('tensor_1', torch.randn(3,3)), ('tensor_2', torch.randn(4,4))]
    >>> defaults = {'sparsity_level': 0.7}
    >>> sparsifier = DerivedDataSparsifier(data_list = data_list, **defaults) # Some sparsifier that inherits BaseDataSparsifier
    >>> new_tensor_to_add = {'name': 'tensor_3', 'data': torch.randn(5,5), 'sparsity_level': 0.3}
    >>> sparsifier.add_data(**new_tensor_to_add)
    >>> # tensor_1 and tensor_2 will have sparsity_level of 0.7 but tensor_3 will have sparsity_level=0.3

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Support for sparsifying data operations on raw torch tensors."


The users can now pass in raw torch tensors and the base class handles all the parametrizations and masking

Example -
    >>> data_list = [('tensor_1', torch.randn(3,3)), ('tensor_2', torch.randn(4,4))]
    >>> defaults = {'sparsity_level': 0.7}
    >>> sparsifier = DerivedDataSparsifier(data_list = data_list, **defaults) # Some sparsifier that inherits BaseDataSparsifier
    >>> new_tensor_to_add = {'name': 'tensor_3', 'data': torch.randn(5,5), 'sparsity_level': 0.3}
    >>> sparsifier.add_data(**new_tensor_to_add)
    >>> # tensor_1 and tensor_2 will have sparsity_level of 0.7 but tensor_3 will have sparsity_level=0.3

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] Support for sparsifying data operations on raw torch tensors."


The users can now pass in raw torch tensors and the base class handles all the parametrizations and masking

Example -
    >>> data_list = [('tensor_1', torch.randn(3,3)), ('tensor_2', torch.randn(4,4))]
    >>> defaults = {'sparsity_level': 0.7}
    >>> sparsifier = DerivedDataSparsifier(data_list = data_list, **defaults) # Some sparsifier that inherits BaseDataSparsifier
    >>> new_tensor_to_add = {'name': 'tensor_3', 'data': torch.randn(5,5), 'sparsity_level': 0.3}
    >>> sparsifier.add_data(**new_tensor_to_add)
    >>> # tensor_1 and tensor_2 will have sparsity_level of 0.7 but tensor_3 will have sparsity_level=0.3

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

[ghstack-poisoned]

macandro96 commented:
@macandro96 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37164121).

Commit:
Update on "[ao][sparsity] Support for sparsifying data operations on raw torch tensors."


The users can now pass in raw torch tensors and the base class handles all the parametrizations and masking

Example -
    >>> data_list = [('tensor_1', torch.randn(3,3)), ('tensor_2', torch.randn(4,4))]
    >>> defaults = {'sparsity_level': 0.7}
    >>> sparsifier = DerivedDataSparsifier(data_list = data_list, **defaults) # Some sparsifier that inherits BaseDataSparsifier
    >>> new_tensor_to_add = {'name': 'tensor_3', 'data': torch.randn(5,5), 'sparsity_level': 0.3}
    >>> sparsifier.add_data(**new_tensor_to_add)
    >>> # tensor_1 and tensor_2 will have sparsity_level of 0.7 but tensor_3 will have sparsity_level=0.3

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

Differential Revision: [D37164121](https://our.internmc.facebook.com/intern/diff/D37164121)

[ghstack-poisoned]

macandro96 commented:
@macandro96 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37164121).

Commit:
Update on "[ao][sparsity] Support for sparsifying data operations on raw torch tensors."


The users can now pass in raw torch tensors and the base class handles all the parametrizations and masking

Example -
    >>> data_list = [('tensor_1', torch.randn(3,3)), ('tensor_2', torch.randn(4,4))]
    >>> defaults = {'sparsity_level': 0.7}
    >>> sparsifier = DerivedDataSparsifier(data_list = data_list, **defaults) # Some sparsifier that inherits BaseDataSparsifier
    >>> new_tensor_to_add = {'name': 'tensor_3', 'data': torch.randn(5,5), 'sparsity_level': 0.3}
    >>> sparsifier.add_data(**new_tensor_to_add)
    >>> # tensor_1 and tensor_2 will have sparsity_level of 0.7 but tensor_3 will have sparsity_level=0.3

Test Plan:
```python test/test_ao_sparsity.py TestBaseDataSparsifier```

Differential Revision: [D37164121](https://our.internmc.facebook.com/intern/diff/D37164121)

[ghstack-poisoned]

macandro96 commented:
@macandro96 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37164121).
